### PR TITLE
Basic tl-tapping consideration for stamina

### PR DESCRIPTION
**Pull request by vunyunt**
Prereqs:
- [x] https://github.com/ppy/osu/pull/20984

The main aim of this PR is to improve difficulty calculation for patterns with extremely fast and long mono patterns. In this case, the player can use more than 2 fingers to press the same keys (known as tl-tapping), making the original assumption of 2 fingers available for each colour false.

### Difficulty Calculator Changes:
- Refactored colour slightly to allow access to colour data in all notes instead of just the first note of each pattern chunk. The detection of first note is moved into colour evaluator.
- Implemented variable available finger count for stamina. Mono notes more than 300ms away from a colour change is considered to have an available finger count of 4.
- Removed convert specific nerfs in difficulty calculation.
- Reduced the minimum note interval for stamina to an unrealistically small value (1ms). This is now only to prevent division by zero or negative values.

### Performance Calculator Changes:
- Introduction of a `rulesetTaiko` bool, which applies previous bonuses strictly to the taiko ruleset as a way to combat easy-to-memorise converts. This is temporary until the reading skill is implemented.
- The HDFL bonus to accuracy has been changed to be able to give 0 bonus, and an increase cap to 1.1x. This was in response to issues discovered within very short maps paired with HDFL giving the same bonus as those with 1000 objects.

---

SR/PP spreadsheets:

**With converts**: https://docs.google.com/spreadsheets/d/1eS4BBq7Ku7oY62nHtOnRBbrnhSTiMphLiHpwyWwEcwA/edit#gid=686072955

**Without converts**: https://docs.google.com/spreadsheets/d/12tejBO5JFGvOfuAKnayHHcWv5MWxOfH591eT6Vhk0h4/edit

As of https://github.com/ppy/osu/pull/20558/commits/9065cb05261fe84519529b4c832f5ecccd7e85f7

### CODE REVIEW

**Review by Lawtrohux — DISMISSED** [review]
looks fine to me. in the future when more skills are worked on we can hope that converts and their respective playstyles don't continue to need special consideration.

**Comment by Lawtrohux** [comment]
@smoogipoo can we get a smoogisheet for this build with converts enabled please?

**Comment by smoogipoo** [comment]
SR/PP sheet added to the description.

**Review by Lawtrohux — DISMISSED** [review]
reapproving as smoogisheet was just to double check unseen values.

what i will say is that some loved converts gain absolute stupid sr, but this solely due to them being impossible maps in the first place, however this issue is past this fix/buff/nerf to things so ill leave that up to discretion having a significant amount of 10* - 40* converts.

**Comment by smoogipoo** [comment]
So just to confirm, it's intended that https://osu.ppy.sh/scores/taiko/151614379 gains 300pp and becomes the new pp record?

**Comment by Lawtrohux** [comment]
> So just to confirm, it's intended that https://osu.ppy.sh/scores/taiko/151614379 gains 300pp and becomes the new pp record?

In regards to other scores gaining, this is justified. Converts were previously fixed nerfed 20% by the nature of the code, and are in a better place now then before. While tweaks still need to be made, this is planned for the next large pr within the coming week containing the rewrite of the rhythm skill.

File: osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Colour/TaikoColourDifficultyPreprocessor.cs [hunk 1]
```diff
@@ -36,14 +37,22 @@ public static void ProcessAndAssign(List<DifficultyHitObject> hitObjects)
...
                     {
                         MonoStreak monoStreak = monoPattern.MonoStreaks[j];
                         monoStreak.Parent = monoPattern;
                         monoStreak.Index = j;
-                        monoStreak.FirstHitObject.Colour.MonoStreak = monoStreak;
+
+                        foreach (var hitObject in monoStreak.HitObjects)
+                        {
+                            hitObject.Colour.MonoStreak = monoStreak;
+                        }
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Colour/TaikoColourDifficultyPreprocessor.cs:55]: Isn't it functionally equivalent to add:
```
hitObject.Colour.AlternatingMonoPattern = monoPattern;
hitObject.Colour.RepeatingPattern = repeatingHitPattern;
```
In this loop? Just to reduce the amount of looping.

File: osu.Game.Rulesets.Taiko/Difficulty/Evaluators/StaminaEvaluator.cs [hunk 2]
```diff
@@ -11,22 +11,43 @@ namespace osu.Game.Rulesets.Taiko.Difficulty.Evaluators
...
+                return 2;
+            }
+
+            if (nextColourChange != null && nextColourChange.StartTime - hitObject.StartTime < 300)
+            {
+                return 2;
+            }
+
+            return 4;
+        }
```
[the remark this answers is not in the document]
Comment (Lawtrohux) [reply osu.Game.Rulesets.Taiko/Difficulty/Evaluators/StaminaEvaluator.cs:46]: mechanically, normal taiko players have 2 fingers available to hit each colour, with the TL playstyle, they use four and above fingers solely to hit the single colour.

2 fingers refers to the normal count, while 4 refers to the new count if the threshold is reached.

**Review by mangomizer — COMMENTED** [review]
Smoogisheet seems to feature mostly converts (which I agree is the main goal of this), and PR does a fair job in tackling this issue.

I would like to see the results for mode-specific maps (we should expect minimal changes) before making any call on this though (can't seem to find many in the smoogisheet).

File: osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Colour/TaikoColourDifficultyPreprocessor.cs (hunk 1 above)
Comment (vunyunt) [reply osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Colour/TaikoColourDifficultyPreprocessor.cs:55]: Yeah it should be, will apply this when I'm back

Comment (vunyunt) [reply osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Colour/TaikoColourDifficultyPreprocessor.cs]: Applied in 122064d03fcdb4fa65dd7ccbbb20eebd4148a33e

**Comment by smoogipoo** [comment]
> I would like to see the results for mode-specific maps

Added to OP.

**Review by mangomizer — DISMISSED** [review]
All good.

**Comment by smoogipoo** [comment]
Updated the spreadsheets in the description because they were generated incorrectly (see https://github.com/ppy/osu-queue-score-statistics/pull/102). Please re-check.

**Review by Lawtrohux — DISMISSED** [review]
lgtm! the new values actually look better as per the accuracy issue

**Review by mangomizer — DISMISSED** [review]
I believe I've raised in the past that it would be useful to run it through a sample of "known" maps as a benchmark, even if just to confirm that the change is minimal. 

If I am right in assuming that maps not showing up is indicative of minimal change (ie. <0.1* SR), approved.

**Comment by smoogipoo** [comment]
> If I am right in assuming that maps not showing up is indicative of minimal change (ie. <0.1* SR), approved.

Correct, it only shows SR differences >0.1*.

**Comment by Lawtrohux** [comment]
I'll just quickly state that while the comment in the detection of rulesets is indeed temporary in regards to specific multipliers. It is planned to be used in a couple of others ways in future PR's, meaning that the related PR's to assist with this are not for a temporary function.

**Review by smoogipoo — DISMISSED** [review]

**Review by smoogipoo — DISMISSED** [review]

**Comment by smoogipoo** [comment]
@peppy or @bdach give this one a once over please.

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs [hunk 3]
```diff
@@ -43,6 +43,9 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
             if (totalSuccessfulHits > 0)
                 effectiveMissCount = Math.Max(1.0, 1000.0 / totalSuccessfulHits) * countMiss;
 
+            // TODO: The detection of rulesets is temporary until the leftover old skills have been reworked.
+            bool taikoSpecificBeatmap = score.BeatmapInfo.Ruleset.OnlineID == 1;
```
Comment (peppy) [inline osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs:47]: Would it be better to call this `isConvert` similar to everywhere else?

Comment (smoogipoo) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs:47]: That's a fair point.

**Review by peppy — APPROVED** [review]
Seems fine code-wise

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**